---
title: "bcRep news"
author: "Julia Bischof <Julia.Bischof@uksh.de>"
date: '`r Sys.Date()`'
output:
  pdf_document:
    toc: yes
    toc_depth: 4
  html_document:
    theme: united
    toc: yes
  word_document: default
---

## Changes in version 1.3.6 (2016-12-16)

### BUG FIXES

     * geneUsage() & compare.geneUsage(): explained parameter 'abundance = "relative"' in 
        help file
     * compare.geneUsage(): eliminated errors
     * clones(): eliminated error for 'useJ = FALSE'

### NEW FEATURES AND FUNCTIONS

     * clones.IDlist(): returns a list containing the sequence ID's (from IMGT table) and the 
        corresponding clone number


## Changes in version 1.3.5 (2016-06-03)

### BUG FIXES

     * geneUsage.distance(): eliminated errors
     * geneUsage() & compare.geneUsage(): eliminated errors

### NEW FEATURES AND FUNCTIONS

     * sequences.mutation.base(): Added function to calculate proportions of silent mutations 
        from nucleotide A to nucleotide B. Now, the nucleotide changes of the mutated position, 
        but also the nucleotide distributions next to the mutated position (-3 to +3) can be 
        analysed.
     * plotSequencesMutationBase(): added function to plot results of sequences.mutation.base()



## Changes in version 1.3.4 (2016-04-21)

### BUG FIXES

     * clones(): eliminated errors


## Changes in version 1.3.3 (2016-04-15)

### BUG FIXES

     * clones.filterSize(): eliminated errors
     * clones.shared(): eliminated errors
     * clones.compareGeneUsage(): eliminated errors

### NEW FEATURES AND FUNCTIONS

     * plotTrueDiversity(): added option to plot mean diversities
     * plotCompareTrueDiversity(): added option to plot mean diversities
     
## Changes in version 1.3.2 (2016-03-17)

### BUG FIXES

     * sequences.geneComb(): eliminated errors
     * compare.geneUsage(): eliminated errors
     

## Changes in version 1.3 (2016-02-25)

### BUG FIXES

     * readIMGT(): eliminated errors
     * clones(): eliminated error message when no clone was found
     * geneUsage(): eliminated error for JH subgroup usage
    

### NEW FEATURES AND FUNCTIONS
      
     * sequences.mutation.AA() added (proportions of amino acid mutations)
     * plotSequencesMutationAA() added
     * sequences.mutation.base() added (proportions of bases around mutations)
     * plotSequencesMutationBase() added


## Changes in version 1.3 (2016-02-15)

### BUG FIXES

     * clones.shared(): reduced computation time and memory; copy number of CDR3 sequences 
        changed
     * sequences.mutation(): added sequence ID's
    

### NEW FEATURES AND FUNCTIONS
      
     * added NEWS
     * sequences.distance() added (distance/dissimilarity measurements on sequence data)
     * geneUsage.distance() added (distance/dissimilarity measurements on gene usage data)
     * dist.PCoA() and plotDistPCoA() added (Principal coordinate analysis on distance data
        and visualization)



## Changes in version 1.2.2 (2015-10-28)

### BUG FIXES

      * plotClonesCopyNumber(): added parameter to plot with and without outliers 
        (clone sizes > 75\% quantile)
      * geneUsage(): Combination of gene usage and functionality didn't work for alleles before
      * sequences.geneComb(): added parallel processing option
      

### NEW FEATURES AND FUNCTIONS
      
     * added Vignette
     * combineIMGT(): combines IMGT output files from different folders
     * clones.giniIndex(): calculates Gini Index of clones
     * clones.filterSize(): filters clones for their size (number, percentage)
     * clones.filterFuntionality(): filters clones for their funtionality
     * clones.filterJunctionFrame(): filters clones for their junction frame usage
     * compare.aaDistribution(), plotCompareAADistribution(): compare amino acid distribution 
        of different samples
     * compare.trueDiversity(), plotCompareTrueDiversity(): compare richness and diversity of 
        different samples
     * compare.geneUsage(), plotCompareGeneUsage(): compare gene usage of different samples 
        (subgroup, gene, allele)
     
## Version 1.0 (2015-10-09)
